packages/react-components/react-carousel-preview/src/components/Carousel/useCarousel.ts

```diff
@@ -105,21 +105,37 @@ export function useCarousel_unstable(props: CarouselProps, ref: React.Ref<HTMLDi
     };
   }, [carouselWalker, store]);
 
-  const selectPageByDirection: CarouselContextValue['selectPageByDirection'] = useEventCallback((event, direction) => {
-    const active = carouselWalker.active();
+  const selectPageByDirection: CarouselContextValue['selectPageByDirection'] = useEventCallback(
+    (event, direction, _loop) => {
+      const active = carouselWalker.active();
 
-    if (!active?.value) {
-      return;
-    }
+      if (!active?.value) {
+        return;
+      }
 
-    const newPage =
-      direction === 'prev' ? carouselWalker.prevPage(active.value) : carouselWalker.nextPage(active.value);
+      let newPage =
+        direction === 'prev' ? carouselWalker.prevPage(active.value) : carouselWalker.nextPage(active.value);
 
-    if (newPage) {
-      setValue(newPage?.value);
-      onValueChange?.(event, { event, type: 'click', value: newPage?.value });
-    }
-  });
+      if (!newPage && _loop) {
```

Comment:
instead of implementing looping, you can update the implementation of the `carouselWalker`. The DOM `TreeWalker` has methods to first and last child

* https://developer.mozilla.org/en-US/docs/Web/API/TreeWalker/firstChild
* https://developer.mozilla.org/en-US/docs/Web/API/TreeWalker/lastChild